src/tls13.rs

```diff
@@ -0,0 +1,313 @@
+                );
+                return Err(rustls::Error::General(custom_error_message));
+            }
+        }
+    }
+
+    fn encrypted_payload_len(&self, payload_len: usize) -> usize {
+        payload_len + 1 + CHACHA_TAG_LENGTH
+    }
+}
+
+/// [`MessageDecrypter`] for ChaCha 1.3
+/// the [`payload`] field that comes from the [`OpaqueMessage`] is structured to include the message which is an arbitrary length,
+/// an encoding type that is 1 byte and then finally the tag which is 16 bytes.
+/// ex : [1, 2, 3, 5, 6, 7, 8, 9, 10, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12 ,13, 14, 15, 16]
+///       ^                       ^   ^  ^                                                   ^
+///            Message (N bytes)   Encoding (1 Byte)              Tag (16 bytes)
+impl MessageDecrypter for Tls13ChaCha20Poly1305 {
+    fn decrypt(&self, mut msg: OpaqueMessage, seq: u64) -> Result<PlainMessage, rustls::Error> {
+        let mut payload = msg.payload_mut();
+        let payload_len = payload.len(); // This length includes the message, encoding, and tag.
+
+        // Ensure that the length is over 16 bytes or there is a decryption error.
+        if payload_len < CHACHA_TAG_LENGTH {
+            return Err(rustls::Error::DecryptError);
+        }
+        let message_length = payload_len - CHACHA_TAG_LENGTH; // getting message length, this includes the message length and the encoding type.
+
+        // Set up needed parameters for ChaCha decrypt
+        let nonce = Nonce::new(&self.iv, seq);
+        let auth_data = make_tls13_aad(payload_len); // The total message including tag and encoding byte must be used for auth data.
+        let mut tag = [0u8; GCM_TAG_LENGTH];
+        tag.copy_from_slice(&payload[message_length..]);
+
+        // Decrypting the payload in place, there is no +1 here since [`message_length`] accounts for the extra byte for encoding type.
+        match chacha20_poly1305_decrypt_in_place(
+            &self.key,
+            &nonce.0,
+            &auth_data,
+            &mut payload[..message_length],
+            &tag,
+        ) {
+            Ok(_) => {
+                payload.truncate(message_length);
+                msg.into_tls13_unpadded_message()
+            }
+            Err(symcrypt_error) => {
+                let custom_error_message = format!(
+                    "SymCryptError: {}",
+                    symcrypt_error.to_string() // Using general error to propagate the SymCrypt error back to the caller
+                );
+                return Err(rustls::Error::General(custom_error_message));
+            }
+        }
+    }
+}
+
+/// GCM for TLS 1.3
+
+/// Tls13Gcm impls [`Tls13AeadAlgorithm`].
+///
+/// [`algo_type`] represents either GCM128 or GCM256 which corresponds to a 16 and 32 byte key respectively.
+pub struct Tls13Gcm {
+    algo_type: AesGcm,
+}
+
+/// Algo types for GCM TLS 1.3 and TLS 1.2
+pub enum AesGcm {
+    Aes128Gcm,
+    Aes256Gcm,
```

Comment:
C-style

Aes128Gcm = 16,
Aes256Gcm = 32,

Reply:
Changed ( chat gpt thinks the the current way is more idomatic though )

Reply:
I agree with chat. The enum type shouldn't be tied directly to the length. Could be multiple enum types with same length.